### PR TITLE
feat: update iOS SDK

### PR DESCRIPTION
**Pull request by duyhungtnn**
# Changes 
- [x] Move `ReasonType` to `internal` scope 
- [x] Send tags when tracking events 
- [x] Fix duplicate event metrics logic 
- [x] Expose BKTUser's properties
- [x] Make BKTError implement LocalizeError 
- [x] Add more test cases to check duplicate metrics event

# Refs 
https://github.com/bucketeer-io/android-client-sdk/pull/64
https://github.com/bucketeer-io/android-client-sdk/pull/68

### CODE REVIEW

File: BucketeerTests/EventInteractorTests.swift [hunk 1]
```diff
@@ -17,10 +20,11 @@ final class EventInteractorTests: XCTestCase {
             eventDao: dao,
             clock: clock,
             idGenerator: idGenerator,
-            logger: logger
+            logger: logger,
+            featureTag: config.featureTag
         )
     }
-
+
```
Comment (duyhungtnn) [inline BucketeerTests/EventInteractorTests.swift]: I added one space to fix the swiftLint warning. So we could see more something like that across this PR

**Comment by duyhungtnn** [comment]
@cre8ivejp  Hi, this PR is ready for review. Please help me to take a look. Let me know if you see any problems. Thank you.

**Comment by cre8ivejp** [comment]
@duyhungtnn

I think we don't need the `// swiftlint:enable type_body_length` [here](https://github.com/bucketeer-io/ios-client-sdk/pull/5/files#diff-d97542f3c89f841f812285facd9fe4d1a1e06d3e9ea2bb903e512f26d93aa3b1R574)
Also, please use `// swiftlint:disable file_length` at the [top of the file](https://github.com/bucketeer-io/ios-client-sdk/pull/5/files#diff-d97542f3c89f841f812285facd9fe4d1a1e06d3e9ea2bb903e512f26d93aa3b1R1) to fix the warning.

Please fix the warning in this [line](https://github.com/bucketeer-io/ios-client-sdk/blob/main/Bucketeer/Sources/Internal/Remote/ApiClientImpl.swift#L14).
It should be as follows.

```swift
    init(
        apiEndpoint: URL,
        apiKey: String,
        ...
```

File: Bucketeer/Sources/Public/BKTError.swift [hunk 2]
```diff
@@ -94,4 +94,68 @@ extension BKTError {
             self = .unknown(message: "Unknown error: \(error)", error: error)
         }
     }
+    
+    /// A localized message describing what error occurred.
+    public var errorDescription: String? {
```
Comment (duyhungtnn) [inline Bucketeer/Sources/Public/BKTError.swift:99]: I saw a problem when updating Flutter SDK, I need many switch cases to read the error.
So that I added the code below :
-> Support SDK's consumer read error messages the easy way. 
@cre8ivejp

File: Bucketeer/Sources/Public/BKTUser.swift [hunk 3]
```diff
@@ -1,8 +1,8 @@
 import Foundation
 
 public struct BKTUser {
-    let id: String
-    let attr: [String: String]
+    public let id: String
+    public let attr: [String: String]
```
Comment (duyhungtnn) [inline Bucketeer/Sources/Public/BKTUser.swift:5]: BKUser's properties are internal access so that the SDK's consumer (Flutter) cannot access the current user data.
It makes BKTClient.getCurrentUser() useless 

@cre8ivejp

File: BucketeerTests/EventInteractorTests.swift [hunk 4]
```diff
@@ -571,4 +571,4 @@ final class EventInteractorTests: XCTestCase {
         wait(for: [expectation], timeout: 1)
     }
 }
-// swiftlint:enable [type_body_length file_length]
+// swiftlint:enable type_body_length file_length
```
Comment (cre8ivejp) [inline BucketeerTests/EventInteractorTests.swift:574]: Just confirming. Does it need to enable at the end of the file?

Comment (duyhungtnn) [reply BucketeerTests/EventInteractorTests.swift:574]: yes, we need to re-enable it at the end of the file, otherwise we will have one more warning like that 

Screenshot 2023-06-19 at 10 46 24

File: Bucketeer/Sources/Public/BKTUser.swift (hunk 3 above)
Comment (cre8ivejp) [reply Bucketeer/Sources/Public/BKTUser.swift:5]: Thanks for fixing it!

File: Bucketeer/Sources/Public/BKTError.swift (hunk 2 above)
Comment (cre8ivejp) [reply Bucketeer/Sources/Public/BKTError.swift:99]: Does this change need to be done also in Android SDK?

Comment (duyhungtnn) [reply Bucketeer/Sources/Public/BKTError.swift:99]: I will check when I am updating the Flutter for Android and send new PR to Android SDK if needed

File: BucketeerTests/MetricsEventUniqueKeyTests.swift [hunk 5]
```diff
@@ -0,0 +1,72 @@
+//
+//  MetricsEventDataTests.swift
+//  BucketeerTests
+//
+//  Created by Ryan Hung Pham on 15/06/2023.
```
Comment (cre8ivejp) [inline BucketeerTests/MetricsEventUniqueKeyTests.swift]: Please delete the header. I'm planning to run a script to add it to all files.

Comment (duyhungtnn) [reply BucketeerTests/MetricsEventUniqueKeyTests.swift]: I removed the header

File: Bucketeer/Sources/Internal/Model/MetricsEventData.swift [hunk 6]
```diff
@@ -1,5 +1,17 @@
 import Foundation
 
+protocol MetricsEventDataProps {
+    var apiId: ApiId { get }
+    var labels: [String: String] { get }
+    var protobufType: String? { get }
```
Comment (cre8ivejp) [inline Bucketeer/Sources/Internal/Model/MetricsEventData.swift:6]: Shouldn't this be required?
I don't remember a case that the `protobufType` will be nil.

Comment (duyhungtnn) [reply Bucketeer/Sources/Internal/Model/MetricsEventData.swift:6]: This protocol follows up with the same current implementation. You could see on each  `metrics_event` 
Yes, I think we should make it non-null as our current test case force it non-null when checking.

Comment (cre8ivejp) [reply Bucketeer/Sources/Internal/Model/MetricsEventData.swift:6]: Understood.
Thanks for fixing it.

**Review by cre8ivejp — APPROVED** [review]
Thank you!
